src/compiler/moduleSpecifiers.ts

```diff
@@ -107,61 +109,63 @@ import {
 interface Preferences {
     readonly relativePreference: RelativePreference;
-    readonly ending: Ending;
+    /**
+     * @param syntaxImpliedNodeFormat Used when the import syntax implies ESM or CJS irrespective of the mode of the file.
+     */
+    getAllowedEndingsInPrefererredOrder(syntaxImpliedNodeFormat?: SourceFile["impliedNodeFormat"]): ModuleSpecifierEnding[];
 }
 
-function getPreferences(host: ModuleSpecifierResolutionHost, { importModuleSpecifierPreference, importModuleSpecifierEnding }: UserPreferences, compilerOptions: CompilerOptions, importingSourceFile: SourceFile): Preferences {
+function getPreferences(
+    { importModuleSpecifierPreference, importModuleSpecifierEnding }: UserPreferences,
+    compilerOptions: CompilerOptions,
+    importingSourceFile: SourceFile,
+    oldImportSpecifier?: string,
+): Preferences {
+    const preferredEnding = getPreferredEnding();
     return {
         relativePreference:
+            oldImportSpecifier !== undefined ? (isExternalModuleNameRelative(oldImportSpecifier) ?
+                RelativePreference.Relative :
+                RelativePreference.NonRelative) :
             importModuleSpecifierPreference === "relative" ? RelativePreference.Relative :
             importModuleSpecifierPreference === "non-relative" ? RelativePreference.NonRelative :
             importModuleSpecifierPreference === "project-relative" ? RelativePreference.ExternalNonRelative :
             RelativePreference.Shortest,
-        ending: getEnding(),
-    };
-    function getEnding(): Ending {
-        switch (importModuleSpecifierEnding) {
-            case "minimal": return Ending.Minimal;
-            case "index": return Ending.Index;
-            case "js": return Ending.JsExtension;
-            default: return usesJsExtensionOnImports(importingSourceFile) || isFormatRequiringExtensions(compilerOptions, importingSourceFile.path, host) ? Ending.JsExtension
-                : getEmitModuleResolutionKind(compilerOptions) !== ModuleResolutionKind.NodeJs ? Ending.Index : Ending.Minimal;
-        }
-    }
-}
-
-function getPreferencesForUpdate(compilerOptions: CompilerOptions, oldImportSpecifier: string, importingSourceFileName: Path, host: ModuleSpecifierResolutionHost): Preferences {
-    return {
-        relativePreference: isExternalModuleNameRelative(oldImportSpecifier) ? RelativePreference.Relative : RelativePreference.NonRelative,
-        ending: hasJSFileExtension(oldImportSpecifier) || isFormatRequiringExtensions(compilerOptions, importingSourceFileName, host) ?
-            Ending.JsExtension :
-            getEmitModuleResolutionKind(compilerOptions) !== ModuleResolutionKind.NodeJs || endsWith(oldImportSpecifier, "index") ? Ending.Index : Ending.Minimal,
+        getAllowedEndingsInPrefererredOrder: syntaxImpliedNodeFormat => {
+            if (syntaxImpliedNodeFormat === ModuleKind.ESNext || (syntaxImpliedNodeFormat ?? importingSourceFile.impliedNodeFormat) === ModuleKind.ESNext) {

```

Comment:
sandersnPending
The second predicate covers the first, so you can get by without it:

```suggestion
            if ((syntaxImpliedNodeFormat ?? importingSourceFile.impliedNodeFormat) === ModuleKind.ESNext) {
```

Reply:
I have literally no idea what I was thinking here. Must have been a sequence of mechanical refactors.